### PR TITLE
Clean/rename FromMultiCl HGCal egamma collections

### PR DESCRIPTION
**Pull request by SohamBhattacharya**
This PR is made in collaboration with @hatakeyamak , and is related to the cleaning discussed in [1].
Tagging @beaudett and @archiron , who are looking at the e-gamma Validation/DQM.
The PR will clean the `*FromMultiCl` HGCal egamma sequences as they've have switched to TICL based reconstruction, and there's no need to run the sim-cluster based sequences for HGCal electrons and photons. So for example:

1. `particleFlowClusterHGCalFromMultiCl` will no longer exist as `particleFlowClusterHGCal` is TICL based.
The sim-cluster based `particleFlowClusterHGCal` is still kept (now called `particleFlowClusterHGCalFromSimCl`, and packed into `_phase2_hgcal_simPFTask` [2]), if it's required by other modules.

2. At RECO: `ecalDrivenGsfElectronsFromMultiCl`/`photonsFromMultiCl` --> `ecalDrivenGsfElectronsHGC`/`photonsHGC`.
The new collections now contain only HGCal electrons and photons due to [3, 4] (earlier they used to contain the barrel too).
The plot shows the eta distributions of `ecalDrivenGsfElectronsHGC` (blue) and `ecalDrivenGsfElectronsFromMultiCl` (red).
hgcal_ele_eta_1

3. At MINIAOD: `slimmedElectronsFromMultiCl`/`slimmedPhotonsFromMultiCl` --> `slimmedElectronsHGC`/`slimmedPhotonsHGC`. These collections should remain the same, as it's just a renaming.
The plot shows the eta distributions of `slimmedElectronsHGC` (blue) and `slimmedElectronsFromMultiCl` (red).
slimmedElectronsHGC_eta

[1] [cms-sw/cmssw#32475 (comment)](https://github.com/cms-sw/cmssw/issues/32475#issuecomment-801167501)
[2] https://github.com/SohamBhattacharya/cmssw/blob/clean_FromMultiCl_12_0_0_pre3/RecoParticleFlow/Configuration/python/RecoParticleFlow_cff.py#L76
[3] https://github.com/SohamBhattacharya/cmssw/blob/clean_FromMultiCl_12_0_0_pre3/RecoEgamma/EgammaElectronProducers/python/ecalDrivenGsfElectronCoresHGC_cff.py#L6
[4] https://github.com/SohamBhattacharya/cmssw/blob/clean_FromMultiCl_12_0_0_pre3/RecoEgamma/EgammaPhotonProducers/python/photonCore_cfi.py#L20

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
@SohamBhattacharya I understand the seen changes are in synch with the PR description, but if you can confirm please:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-07-11-2300+0e3b41/44003/dqm-histo-comparison-summary.html
Thanks!

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,5 @@ __init__.py
 #*#
 *~
 *.pb
+*.root
+*replace.py
```
Comment (makortel) [inline .gitignore:9]: Please remove these changes (also from the commit history). Additoin of `*.root` could be discussed, but it would be better done as a separate PR.

**Comment by SohamBhattacharya** [comment]
> @SohamBhattacharya I understand the seen changes are in synch with the PR description, but if you can confirm please:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-07-11-2300+0e3b41/44003/dqm-histo-comparison-summary.html
> Thanks!

@jfernan2 I went through the change log, and the changes are related to the `*FromMultiCl`, `particleFlowClusterHGCal`, and `particleFlowSuperClusterHGCal` collections, which is expected.

**Comment by slava77** [comment]
>     3\. The plot shows the eta distributions of `slimmedElectronsHGC` (blue) and `slimmedElectronsFromMultiCl` (red).
https://user-images.githubusercontent.com/15279834/125169147-21ef8500-e1a9-11eb-96dc-fef72e198d37.png

I'm curious about the entry around eta of 0.8. If this is HGCAL, how did it get created?

File: RecoEcal/Configuration/python/RecoEcal_EventContent_cff.py [hunk 2]
```diff
@@ -66,7 +64,7 @@
 )
 RecoEcalRECO.outputCommands.extend(RecoEcalAOD.outputCommands)
 _phase2_hgcal_scCommands = ['keep *_particleFlowSuperClusterHGCal_*_*',
-                            'keep *_particleFlowSuperClusterHGCalFromMultiCl_*_*']
+                            'keep *_particleFlowSuperClusterHGCal_*_*']
```
Comment (slava77) [inline RecoEcal/Configuration/python/RecoEcal_EventContent_cff.py:67]:
```suggestion
                            ]
```
this is a duplicate now (please edit to have just one line)

File: RecoLocalCalo/Configuration/python/hgcalLocalReco_cff.py [hunk 3]
```diff
@@ -19,7 +19,7 @@
                                        hgcalMultiClusters,
                                        particleFlowRecHitHGC,
                                        particleFlowClusterHGCal,
-                                       particleFlowClusterHGCalFromMultiCl )
+                                       particleFlowClusterHGCal )
```
Comment (slava77) [inline RecoLocalCalo/Configuration/python/hgcalLocalReco_cff.py:22]:
```suggestion
                                        )
```
this is a duplicate

File: TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 from Configuration.Eras.Modifier_phase2_hgcal_cff import phase2_hgcal
 phase2_hgcal.toModify( TrackingToolsAOD,
-                       outputCommands = TrackingToolsAOD.outputCommands + ['keep recoGsfTracks_electronGsfTracksFromMultiCl_*_*'])
+                       outputCommands = TrackingToolsAOD.outputCommands + ['keep recoGsfTracks_electronGsfTracks_*_*'])
```
Comment (slava77) [inline TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py:12]: this part is not needed anymore (`recoGsfTracks_electronGsfTracks_` are a part of the standard content)

File: TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py [hunk 5]
```diff
@@ -23,11 +23,11 @@
...
-        'keep TrackingRecHitsOwned_electronGsfTracksFromMultiCl_*_*',
-        'keep *_electronMergedSeedsFromMultiCl_*_*'
+        'keep recoGsfTracks_electronGsfTracks_*_*',
+        'keep recoGsfTrackExtras_electronGsfTracks_*_*',
+        'keep recoTrackExtras_electronGsfTracks_*_*',
+        'keep TrackingRecHitsOwned_electronGsfTracks_*_*',
+        'keep *_electronMergedSeeds_*_*'
 ]
 phase2_hgcal.toModify( TrackingToolsRECO,
                        outputCommands = TrackingToolsRECO.outputCommands + _phase2_hgcal_TrackingRECO_tokeep)
```
Comment (slava77) [inline TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py:33]: this part is not needed anymore (everything is already in `TrackingToolsRECO`)

**Review by slava77 — COMMENTED** [review]
there are a few places with apparently redundant modules and product keep statements.
Please check; I'm not fully sure if these redundant parts are incorrect replacements or indeed redundant.

**Comment by SohamBhattacharya** [comment]
> > ```
> > 3\. The plot shows the eta distributions of `slimmedElectronsHGC` (blue) and `slimmedElectronsFromMultiCl` (red).
> > ```
> 
> https://user-images.githubusercontent.com/15279834/125169147-21ef8500-e1a9-11eb-96dc-fef72e198d37.png
> 
> I'm curious about the entry around eta of 0.8. If this is HGCAL, how did it get created?

@slava77 The plots show the electron eta which uses track information. In general it's very close to the supercluster eta, but there can be rare occasions where they differ.
I have attached the plot of the slimmedElectron supercluster eta, and now the entry at 0.8 is no longer there.

slimmedElectronsHGC_sc_eta-1

File: RecoEcal/Configuration/python/RecoEcal_EventContent_cff.py (hunk 2 above)
Comment (SohamBhattacharya) [reply RecoEcal/Configuration/python/RecoEcal_EventContent_cff.py]: Done.

File: RecoLocalCalo/Configuration/python/hgcalLocalReco_cff.py (hunk 3 above)
Comment (SohamBhattacharya) [reply RecoLocalCalo/Configuration/python/hgcalLocalReco_cff.py:21]: Done.

File: TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py (hunk 4 above)
Comment (SohamBhattacharya) [reply TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py]: Done.

File: TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py (hunk 5 above)
Comment (SohamBhattacharya) [reply TrackingTools/Configuration/python/TrackingTools_EventContent_cff.py]: Done.

**Comment by jfernan2** [comment]
please test

**Comment by SohamBhattacharya** [comment]
> -1
> 
> **Failed Tests**: UnitTests
> **Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e3b41/16978/summary.html
> **COMMIT**: [0f17d95](https://github.com/cms-sw/cmssw/commit/0f17d95db53f1f861b52e71c598d3c55c44dfd15)
> **CMSSW**: CMSSW_12_0_X_2021-07-18-2300/slc7_amd64_gcc900
> **User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34439/16978/install.sh` to create a dev area with all the needed externals and cmssw changes.
> ## Unit Tests
> 
> I found errors in the following unit tests:
> 
> ---> test TestDQMServicesDemo had ERRORS
> 
> ## Comparison Summary
> 
> @slava77 comparisons for the following workflows were not done due to missing matrix map:
> 
>     * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-0e3b41/11634.912_TTbar_14TeV+2021_DD4hepDB+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA
> 
> 
> **Summary**:
> 
>     * No significant changes to the logs found
> 
>     * ROOTFileChecks: Some differences in event products or their sizes found
> 
>     * Reco comparison results: 2089 differences found in the comparisons
> 
>     * DQMHistoTests: Total files compared: 39
> 
>     * DQMHistoTests: Total histograms compared: 2996268
> 
>     * DQMHistoTests: Total failures: 3149
> 
>     * DQMHistoTests: Total nulls: 1
> 
>     * DQMHistoTests: Total successes: [2993096](https://github.com/cms-sw/cmssw/commit/2993096fc87659ff8d33169d447904da5de54a2d)
> 
>     * DQMHistoTests: Total skipped: 22
> 
>     * DQMHistoTests: Total Missing objects: 0
> 
>     * DQMHistoSizes: Histogram memory added: -0.004 KiB( 38 files compared)
> 
>     * DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
> 
>     * Checked 165 log files, 37 edm output root files, 39 DQM output files
> 
>     * [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-07-18-2300+0e3b41/44212/triggerResults): no differences found

I'm not sure why the unit test failed. Is it something on my end?

**Comment by jfernan2** [comment]
Failure is not related to this PR, but to other,  as reported in #34520
should be fixed by #34532 which is now merged

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by slava77** [comment]
assign upgrade

looking at the validation plots, do I understand correctly that the default validation now covers only HGCAL
with this PR (red) only endcaps show up
image

where are the barrel electrons monitored/validated?

if the barrel electron validation is now disabled, this seems like a step backward.
Is the assumption that the barrel will be validated privately now?

File: RecoEgamma/EgammaPhotonProducers/python/photonCore_cfi.py [hunk 6]
```diff
@@ -16,9 +16,10 @@
 #    MVA_weights_location = cms.string('RecoEgamma/EgammaTools/data/TMVAnalysis_Likelihood.weights.txt')
 )
 
-photonCoreFromMultiCl = photonCore.clone(
-    scIslandEndcapProducer = 'particleFlowSuperClusterHGCalFromMultiCl',
-    pixelSeedProducer = 'electronMergedSeedsFromMultiCl'
+photonCoreHGC = photonCore.clone(
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/python/photonCore_cfi.py:19]: logs now show
```
%MSG-e PhotonCoreProducer:  PhotonCoreProducer:photonCoreHGC  20-Jul-2021 10:58:48 CEST Run: 1 Event: 2
Error! Can't get the scHybridBarrelProducer
```

is there a configuration flag to silence it?

**Comment by slava77** [comment]
Here are some notes as a summary of the current status of the reco review
- the last updates in the PR addressed the previous review 
- I checked the saved products and it looks like the new collections are as expected
- also, since it's not in the default tests, I checked that a new AODSIM is still OK to run PAT on it using 34834.999

Pending (new; unfortunately noticed only now):
- logError appearing in `PhotonCoreProducer:photonCoreHGC`
- [a bit outside of reco] I'd like to get a clarification about the barrel electron monitoring/validation

**Comment by beaudett** [comment]
Hi @slava77 

about [your question/comment here](https://github.com/cms-sw/cmssw/pull/34439#issuecomment-885308476). Until recently, all the electron DQM modules from the simulation validation were looking at looking at one collection at a time. With @archiron, we started to enable the production of one set of histograms from two different barrel and endcap collections with [this PR](https://github.com/cms-sw/cmssw/pull/33421) in the miniAOD.
The ElectronFakeValidator was showing electrons in barrel and endcaps because it was reading the electrons from a merged collection, while now barrel and endcap electrons are in two different collections. So what you see is completely expected and I would not quality it as a step backward as the barrel electrons in this merged collections are not optimally reconstructed: they use a pre-PF algorithm and not benefit from the recent energy scale corrections. See the red curve in [this plot](https://cms-egamma.web.cern.ch/validation/Electrons/Releases/12_0_0_pre1_PHASE2_DQM_std/FullvsFull_CMSSW_12_0_0_pre1/RECO-miniAOD_ZEE_14/gifs/h_ele_PoPtrueVsEta_pfx.gif)
With @archiron, it is in our (post-vacation) plans to enable the reading of two collections in all electron validation modules.

**Comment by slava77** [comment]
@beaudett 
thank you for clarifying the situation. You reminded me that the full eta range validation is available in miniAOD. If the plan is to eventually monitor the reco collections in a combination, then for this part this PR is good to go.

So, the remaining point for this PR is to cleanup the logError appearing in `PhotonCoreProducer:photonCoreHGC`

**Comment by SohamBhattacharya** [comment]
> Here are some notes as a summary of the current status of the reco review
> 
>     * the last updates in the PR addressed the previous review
> 
>     * I checked the saved products and it looks like the new collections are as expected
> 
>     * also, since it's not in the default tests, I checked that a new AODSIM is still OK to run PAT on it using 34834.999
> 
> 
> Pending (new; unfortunately noticed only now):
> 
>     * logError appearing in `PhotonCoreProducer:photonCoreHGC`
> 
>     * [a bit outside of reco] I'd like to get a clarification about the barrel electron monitoring/validation

@slava77 thank you for the summary.
An hgcalOnly flag has now been added [1] to handle the case where the barrel collection is not provided, without any LogError.

[1] https://github.com/SohamBhattacharya/cmssw/blob/clean_FromMultiCl_12_0_0_pre3/RecoEgamma/EgammaPhotonProducers/python/photonCore_cfi.py#L16

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@cmsbuild please test
(Let retry, hoping we can get rid of the unrelated relval-input failures)

File: RecoEgamma/EgammaPhotonProducers/src/PhotonCoreProducer.cc [hunk 7]
```diff
@@ -74,6 +75,7 @@ PhotonCoreProducer::PhotonCoreProducer(const edm::ParameterSet& config)
   pixelSeedProducer_ = consumes<reco::ElectronSeedCollection>(conf_.getParameter<edm::InputTag>("pixelSeedProducer"));
   minSCEt_ = conf_.getParameter<double>("minSCEt");
   risolveAmbiguity_ = conf_.getParameter<bool>("risolveConversionAmbiguity");
+  hgcalOnly_ = conf_.getParameter<bool>("hgcalOnly");
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/src/PhotonCoreProducer.cc:78]: there is nothing else in this module that suggests that HGCAL is used.
The separation is into Barrel and Endcap.
```suggestion
  endcapOnly_ = conf_.getParameter<bool>("endcapOnly");
```

Comment (SohamBhattacharya) [reply RecoEgamma/EgammaPhotonProducers/src/PhotonCoreProducer.cc]: @slava77 you are right. The change is now implemented.

**Comment by perrotta** [comment]
@cms-sw/analysis-l2 do you have any comment on this PR?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34439	f649e65 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences in the phase-2 workflows and correspond to removal/renaming of the collections and also the change of the reco electron collection `HGC` to be solely covering the HGCAL.

The following notes/summaries may be useful:
- https://github.com/cms-sw/cmssw/pull/34439#issuecomment-885319562
- https://github.com/cms-sw/cmssw/pull/34439#issuecomment-886077635

File: RecoTauTag/RecoTau/python/PFRecoTauDiscriminationAgainstElectronMVA6Phase2_cff.py [hunk 8]
```diff
@@ -1,18 +1,18 @@
...
 
 # HGCal electron stuff
-from RecoEgamma.EgammaTools.cleanedEcalDrivenGsfElectronsFromMultiCl_cfi import cleanedEcalDrivenGsfElectronsFromMultiCl
+from RecoEgamma.EgammaTools.cleanedEcalDrivenGsfElectronsHGC_cfi import cleanedEcalDrivenGsfElectronsHGC
 from RecoEgamma.EgammaTools.hgcalElectronIDValueMap_cff import hgcalElectronIDValueMap
 # HGCal electrons cleaned against duplicates and electrons in barrel (pt>10GeV)
 # TauValElectronSelector defined Validation/RecoTau/plugins/Selectors.cc;
 # is there a more intuitive place where such a selector is defined?
-cleanedEcalDrivenGsfElectronsFromMultiClNoEB = cms.EDFilter('TauValElectronSelector',
+cleanedEcalDrivenGsfElectronsNoEB = cms.EDFilter('TauValElectronSelector',
```
Comment (srimanob) [inline RecoTauTag/RecoTau/python/PFRecoTauDiscriminationAgainstElectronMVA6Phase2_cff.py:9]: Should it not be named as "cleanedEcalDrivenGsfElectronsHGCNoEB" for consistency?

Comment (SohamBhattacharya) [reply RecoTauTag/RecoTau/python/PFRecoTauDiscriminationAgainstElectronMVA6Phase2_cff.py]: @srimanob you're right, I had missed that. Fixed in the latest commit.

**Comment by srimanob** [comment]
Please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34439  7900f04
- since the last reco signoff https://github.com/cms-sw/cmssw/pull/34439#issuecomment-887063624 there was a somewhat technical renaming of an intermediate data product in PFRecoTauDiscriminationAgainstElectronMVA6Phase2_cff.py to improve consistency
- jenkins tests pass and comparisons with the baseline show, as in the previous tests, differences in the phase-2 workflows as expected (IIUC, the last change is not expected to introduce any changes to the final outputs)

**Comment by jfernan2** [comment]
+1
DQM Changes expected

**Comment by srimanob** [comment]
Please test

Give it another try.

File: DQMOffline/EGamma/python/egammaDQMOffline_cff.py [hunk 9]
```diff
@@ -36,7 +36,7 @@
 stdPhotonAnalysisHGCal = stdPhotonAnalysis.clone()
 stdPhotonAnalysisHGCal.ComponentName = 'stdPhotonAnalyzerHGCalFromMultiCl'
```
Comment (srimanob) [inline DQMOffline/EGamma/python/egammaDQMOffline_cff.py:37]: Not sure if I understand if you would like to keep the ComponentName and analyzerName as before, `stdPhotonAnalyzerHGCalFromMultiCl`. 

This comment includes 
https://github.com/cms-sw/cmssw/blob/master/DQMOffline/EGamma/python/photonOfflineDQMClient_cff.py#L13

Comment (srimanob) [reply DQMOffline/EGamma/python/egammaDQMOffline_cff.py:37]: Sorry for late comment @SohamBhattacharya 
I just look through the change in the histogram and have a question.

Comment (SohamBhattacharya) [reply DQMOffline/EGamma/python/egammaDQMOffline_cff.py:37]: @srimanob I had left some of the names in the DQM modules unchanged as the DQM updates are to follow soon [1]. However, it's true that it would be good to update the names to be consistent with the new naming convention.
After consulting our DQM/Validation experts (@beaudett  and @archiron), I have updated the names in the latest commit.

[1] https://github.com/cms-sw/cmssw/pull/34439#issuecomment-886063450

**Comment by srimanob** [comment]
Please test

File: DQMOffline/EGamma/python/egammaDQMOffline_cff.py (hunk 9 above)
Comment (srimanob) [reply DQMOffline/EGamma/python/egammaDQMOffline_cff.py]: Thanks @SohamBhattacharya

Comment (Dr15Jones) [reply DQMOffline/EGamma/python/egammaDQMOffline_cff.py]: It would be better to change this all to be
```python
stdPhotonAnalysisHGCal = stdPhotonAnalysis.clone(
  ComponentName = 'stdPhotonAnalyzerHGCalFromMultiCl',
  analyzerName = 'stdPhotonAnalyzerHGCalFromMultiCl',
  phoProducer = 'photonsHGC',
  isolationStrength = 2,
  etaMin = -3.0,
  etaMax = 3.0,
  maxPhoEta = 3.0
)
```

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34439	d00db96 
- updates since the last reco signoff were in DQMOffline/EGamma

**Comment by slava77** [comment]
@cms-sw/dqm-l2 and @cms-sw/upgrade-l2 
please clarify if you have pending comments to this PR.

**Comment by srimanob** [comment]
Hi @slava77 @SohamBhattacharya 
I am waiting if @SohamBhattacharya will follow a suggestion in https://github.com/cms-sw/cmssw/pull/34439#discussion_r681104177

**Comment by SohamBhattacharya** [comment]
> Hi @slava77 @SohamBhattacharya
> I am waiting if @SohamBhattacharya will follow a suggestion in [#34439 (comment)](https://github.com/cms-sw/cmssw/pull/34439#discussion_r681104177)

Sorry, I had missed that comment. This has now been implemented in the latest commit.

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34439	a361668 
- updates since the last reco signoff were in DQMOffline/EGamma

**Comment by srimanob** [comment]
@cmsbuild please test

Give it another try on relvals input test

**Comment by srimanob** [comment]
+Upgrade

As the PR title, this PR is to clean and rename HGCal EGamma collections. Change in collection name in DQM is expected.

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1
- All comments received during the review have been addressed, results are in agreement with the expectations

**Comment by perrotta** [comment]
merge
